### DISCUSSION TITLE
Access route path parameters in middleware

### DISCUSSION THREAD

**Discussion by Sheemap** [aws-powertools/powertools-lambda-python#4028] [Q&A] [answered]
Hey all!

I've recently started using the new(ish) middleware that can be used both globally and on a route. So far its been fantastic! Really enjoy how yall have designed it.

One thing that Im not sure is covered is being able to pull out path params in the middleware. In my use case we are trying to authorize a user to act on a resource specified via path, for example with the route: `DELETE /v1/documents/<document_id>`

We want to pull out the document_id, and validate that the current user does in fact have delete permissions on it. Id love to build this as a middleware, rather than copying it onto every method that needs this functionality.

If we pass it as a query string we can use `app.current_event.get_query_string_value()`, is it possible we could do something similar for path params?

Thanks!

**Comment by leandrodamascena** [comment]
Hey @Sheemap! 🚀 Thanks a ton for the fantastic feedback on this feature! Middleware is proving to be a great feature, streamlining processes, and simplifying complexities for our customers.

If you have a route defined as `/v1/documents/<document_id>` and NOT `{proxy+}`, so you can access the `path_parameters` using this:

### code
```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.event_handler import APIGatewayRestResolver, Response
from aws_lambda_powertools.event_handler.middlewares import NextMiddleware

app = APIGatewayRestResolver()
logger = Logger()


def inject_correlation_id(app: APIGatewayRestResolver, next_middleware: NextMiddleware) -> Response:
    print(app.current_event.path_parameters)


@app.get("/v1/documents/<document_id>", middlewares=[inject_correlation_id])  
def get_todos(document_id: int):
    return document_id


@logger.inject_lambda_context(correlation_id_path='headers."x-correlation-id"')  
def lambda_handler(event, context):
    return app.resolve(event, context)
```
### output
```sh
START RequestId: 679e1f01-db51-4b3d-b405-64c805e1a146 Version: $LATEST
{'document_id': '1223434556'}
END RequestId: 679e1f01-db51-4b3d-b405-64c805e1a146
REPORT RequestId: 679e1f01-db51-4b3d-b405-64c805e1a146	Init Duration: 0.17 ms	Duration: 304.89 ms	Billed Duration: 305 ms	Memory Size: 128 MB	Max Memory Used: 128 MB	

2024-03-27 19:17:58 127.0.0.1 - - [27/Mar/2024 19:17:58] "GET /v1/documents/1223434556 HTTP/1.1" 200 -
```

Please let me know if I can assist you further in resolving the issue.

**Reply by Sheemap** [reply]
Hey thanks for your response!

Currently we have a monolithic function where we _do_ scope everything with `{proxy+}`. Is there any way to pull out the id with powertools in this case? We've had some terraform-related weirdness when using the non-wildcarded routes, and I would love to not have to revisit that :grin:

**Reply by leandrodamascena** [reply, marked as answer]
If you are using Proxy then your `path_parameters` will be the full path and it is difficult to extract information, I agree.

If this is blocking you and is a critical requirement for your system, there is an alternative method to obtain this information more easily. However, it's worth noting that this approach is not considered best practice and may be subject to internal changes in the code with potential breaking changes.

We need to inject the route arguments in the context because we need to use it when invoking `next_middleware` or the next route, so, you can access the route path parameters by using `_route_args` variable.

_route_args = {'_route': <aws_lambda_powertools.event_handler.api_gateway.Route object at 0x7f780a7f2500>, '_path': '/v1/documents/1223434556', '_route_args': {'document_id': '1223434556'}}

```python
def inject_correlation_id(app: APIGatewayRestResolver, next_middleware: NextMiddleware) -> Response:
    print(app.context.get("_route_args"))


@app.get("/v1/documents/<document_id>", middlewares=[inject_correlation_id])  
def get_todos(document_id: int):
    return document_id
```

I think we should create a new public property called `route_args` and return this to the Middleware. Just like you, others may have this need. I'll think about this.

**Reply by Sheemap** [reply]
Great, yeah that makes sense. Luckily this middleware isnt critical, we can accomplish it in slightly less optimal ways without using the private properties.

Having a public `route_args` property sounds perfect, if you do end up going that route, would you mind mentioning this discussion on the issue/PR? I would love to be notified of its progress.

**Reply by leandrodamascena** [reply]
I'll need a few days to complete some other tasks, but I'll definitely look into this afterward.

If we decide to proceed with implementing this new property, I'll let you know and if you're interested, you would send a pull request to implement it :rocket:.

Pls let me know if I can help with anything else @Sheemap! Thanks

**Reply by Sheemap** [reply]
Great! Sounds good, I'm definitely open to helping contribute a PR.

Thanks a ton for your quick and helpful responses! Have a good one :)

**Comment by leandrodamascena** [comment]
I'm closing this discussion as resolved. Please reopen if you need.
